python/pyarrow/tests/test_pandas.py

```diff
@@ -4411,6 +4412,31 @@ def test_to_pandas_extension_dtypes_mapping():
     assert isinstance(result['a'].dtype, pd.PeriodDtype)
 
 
+
+def test_to_pandas_extension_dtypes_mapping_complex_type():
+    pa_type = pa.struct(
+        [
+            pa.field("bar", pa.bool_(), nullable=False),
+            pa.field("baz", pa.float32(), nullable=True),
+        ],
+    )
+    pd_type = pd.ArrowDtype(pa_type)
+    schema = pa.schema([pa.field("foo", pa_type)])
+    df0 = pd.DataFrame(
+        [
+            {"foo": {"bar": True, "baz": np.float32(1)}},
+            {"foo": {"bar": True, "baz": None}},
+        ],
+    ).astype({"foo": pd_type})
+
+    # Round trip df0 into df1
+    with io.BytesIO() as stream:
+        df0.to_parquet(stream, schema=schema)
+        stream.seek(0)
+        df1 = pd.read_parquet(stream, dtype_backend="pyarrow")
```

Comment:
You might not need the roundtrip to parquet, but a `table = pa.table(df); result = table.to_pandas(types_mapper=pd.ArrowDtype)` should be sufficient to test this?

I know this doesn't test exactly `pd.read_parquet` in its entirety, but it should test the relevant part on the pyarrow side, and an actual pd.read_parquet test can still be added to pandas

Reply:
The error only gets thrown once the pandas metadata is added to the table. That's why I have used a round-trip test. Is there another way to generate that metadata and set it on the table before calling to_pandas?

Reply:
Addressed in https://github.com/apache/arrow/pull/44720/commits/1c23076fb37f8903762c7416db12e8e453e23366